File: src/main/java/cspfj/constraint/ResidueManagerMap.java
package cspfj.constraint;

import static com.google.common.collect.Maps.newHashMap;

import java.util.Arrays;
import java.util.Map;

public final class ResidueManagerMap implements ResidueManager {

    private final Map<Integer, int[]>[] last;

    @SuppressWarnings("unchecked")
    public ResidueManagerMap(final int arity) {
        last = (Map<Integer, int[]>[]) new Map<?, ?>[arity];
        for (int i = arity; --i >= 0;) {
            last[i] = newHashMap();
        }
    }

    public int[] getResidue(final int position, final int index) {
        return last[position].get(index);
    }

    public void updateResidue(final int[] residue) {
        for (int position = last.length; --position >= 0;) {
            last[position].put(residue[position], residue);
        }
    }

    public void remove(final int[] residue) {
        for (int position = last.length; --position >= 0;) {
            if (Arrays.equals(residue, last[position].get(residue[position]))) {
                last[position].remove(residue[position]);
            }
        }
    }
}

File: src/main/scala/cspfj/constraint/ResidueManagerMap.scala
package cspfj.constraint;
import java.util.Arrays
import scala.annotation.tailrec

final class ResidueManagerMap(arity: Int) {

  val last = (0 until arity).map { _ => Map[Int, Array[Int]]().withDefaultValue(null) } toArray

  def getResidue(position: Int, index: Int) = last(position)(index)

  def updateResidue(residue: Array[Int], i: Int = arity - 1) {
    if (i >= 0) {
      last(i) += residue(i) -> residue
      updateResidue(residue, i - 1)
    }
  }

  def remove(residue: Array[Int]) {
    for (i <- residue.indices) {
      if (Arrays.equals(residue, last(i)(residue(i)))) {
        last(i) -= residue(i)
      }
    }
  }
}
